### PR TITLE
CounterSimulator: Add ability to count multi-qubit operations

### PR DESCRIPTION
**Pull request by tcNickolas**
In some tasks, we need to restrict usage of certain multi-qubit operations. To do this, `CounterSimulator` needs to be able to track the number of qubit parameters the operation received.

This will be part of the fix for #74, among its other uses. The second part of the fix will come after we update the Nuget package with this change and enable the kata Notebooks to use it.

### CODE REVIEW

**Comment by tcNickolas** [comment]
The code in `JointMeasurements` kata that exercises the new feature looks like this:

    // A pair of operations to fine-tune universal CounterSimulator
    // for the purposes of the last two tasks: prohibiting all multi-qubit operations,
    // except the two that can be used for solving this task
    operation ResetMultiQubitNonMeasurementOpCount () : Unit {
        ResetMultiQubitOpCount();
        ResetOracleCallsCount();
    }

    operation GetMultiQubitNonMeasurementOpCount () : Int {
        return GetMultiQubitOpCount() - GetOracleCallsCount(Measure) - GetOracleCallsCount(MeasureAllZ);
    }

Then I require that `GetMultiQubitNonMeasurementOpCount() <= 1` in the tasks that prohibit use of non-measurement multiqubit operations. The `1` accounts for calling the task that is being tested itself.

File: utilities/Common/CounterSimulator.cs [hunk 1]
```diff
@@ -193,5 +205,60 @@ public GetMaxQubitCountImpl(CounterSimulator m) : base(m)
...
+        #region Count multi-qubit operations (except Measure)
+        public void CountMultiQubitOperations(ICallable op, IApplyData data)
+        {
+            int nQubits = 0;
+            using (IEnumerator<Qubit> enumerator = data?.Qubits?.GetEnumerator()) {
+                if (enumerator is null) {
+                    // the operation doesn't have qubit parameters
+                    return;
+                }
+                while (enumerator.MoveNext() && nQubits < 2)
```
Comment (anpaz-msft) [inline utilities/Common/CounterSimulator.cs]: I would do `nQubits = enumerator.Take(2).Count();`.

File: utilities/Common/CounterSimulator.cs [hunk 2]
```diff
@@ -193,5 +205,60 @@ public GetMaxQubitCountImpl(CounterSimulator m) : base(m)
...
+            }
+            if (nQubits >= 2) {
+                _multiQubitOperations++;
+            }
+        }
+
+        /// <summary>
+        /// Custom operation to reset the numbers of multi-qubit operations.
+        /// </summary>
+        public class ResetMultiQubitOpCountImpl : ResetMultiQubitOpCount
```
Comment (anpaz-msft) [inline utilities/Common/CounterSimulator.cs]: If you go the always count multi-qubit gates, this is probably not needed.

File: utilities/Common/CounterSimulator.cs (hunk 1 above)
Comment (tcNickolas) [reply utilities/Common/CounterSimulator.cs]: `Take` seems to be a method of LINQ `IEnumerable`, while `enumerator` is `IEnumerator`, and I can't find a neat way to cast `IEnumerator` to `IEnumerable`. I'm not sure the benefits of using `.Take` outweigh the hassle of adding LINQ to the dependencies and casting the enumerables.

**Pull request merged**